File: 1001-2000/1209. Remove All Adjacent Duplicates in String II.java
// Medium
// Stack, String
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/remove-all-adjacent-duplicates-in-string-ii/

import java.util.Stack;

class Solution {
    public String removeDuplicates(String s, int k) {
        // Stack save [character, continuous appear time]
        Stack<int[]> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (!stack.isEmpty() && stack.peek()[0] == c) {
                stack.peek()[1]++;

                if (stack.peek()[1] == k) {
                    stack.pop();
                }
            } else {
                stack.push(new int[] { c, 1 });
            }
        }
        
        StringBuilder sb = new StringBuilder();
        for (int[] pair : stack) {
            char c = (char) pair[0];
            int cnt = pair[1];
            for (int i = 0; i < cnt; i++) {
                sb.append(c);
            }
        }
        return sb.toString();
    }
}

File: 1001-2000/1209. Remove All Adjacent Duplicates in String II.py
# Medium
# Stack
# Time:O(n), Space:O(n)
# https://leetcode.cn/problems/remove-all-adjacent-duplicates-in-string-ii/

class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        stack = []  # (char, count)

        for char in s:
            if stack and stack[-1][0] == char:
                stack[-1] = (char, stack[-1][1] + 1)

                if stack[-1][1] == k:
                    stack.pop()
            else:
                stack.append((char, 1))

        res = []
        for char, count in stack:
            res.append(char * count)

        return "".join(res)
